src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs

```diff
@@ -7612,5 +7612,477 @@ .maxstack  1
+  .maxstack  2
+  .locals init (T V_0, //t
+                object V_1,
+                T V_2,
+                T V_3)
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  stloc.1
+  IL_0003:  ldloc.1
+  IL_0004:  isinst     ""T""
+  IL_0009:  ldnull
+  IL_000a:  cgt.un
+  IL_000c:  dup
+  IL_000d:  brtrue.s   IL_001a
+  IL_000f:  ldloca.s   V_2
+  IL_0011:  initobj    ""T""
+  IL_0017:  ldloc.2
+  IL_0018:  br.s       IL_0020
+  IL_001a:  ldloc.1
+  IL_001b:  unbox.any  ""T""
+  IL_0020:  stloc.0
+  IL_0021:  brtrue.s   IL_002e
+  IL_0023:  ldloca.s   V_2
+  IL_0025:  initobj    ""T""
+  IL_002b:  ldloc.2
+  IL_002c:  br.s       IL_002f
+  IL_002e:  ldloc.0
+  IL_002f:  stloc.3
+  IL_0030:  br.s       IL_0032
+  IL_0032:  ldloc.3
+  IL_0033:  ret
+}"
+            );
+            compVerifier.VerifyIL("Program.M2<T>",
+@"{
+  // Code size       82 (0x52)
+  .maxstack  2
+  .locals init (System.ValueType V_0,
+                T V_1,
+                System.ValueType V_2,
+                T V_3, //t
+                System.ValueType V_4,
+                object V_5,
+                T V_6,
+                T V_7)
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  stloc.2
+  IL_0003:  ldloc.2
+  IL_0004:  stloc.s    V_4
+  IL_0006:  ldloc.s    V_4
+  IL_0008:  stloc.0
+  IL_0009:  ldloc.0
+  IL_000a:  brtrue.s   IL_000e
+  IL_000c:  br.s       IL_0036
+  IL_000e:  ldloc.0
+  IL_000f:  stloc.s    V_5
+  IL_0011:  ldloc.s    V_5
+  IL_0013:  isinst     ""T""
+  IL_0018:  ldnull
+  IL_0019:  cgt.un
+  IL_001b:  dup
+  IL_001c:  brtrue.s   IL_002a
+  IL_001e:  ldloca.s   V_6
+  IL_0020:  initobj    ""T""
+  IL_0026:  ldloc.s    V_6
+  IL_0028:  br.s       IL_0031
+  IL_002a:  ldloc.s    V_5
+  IL_002c:  unbox.any  ""T""
+  IL_0031:  stloc.1
+  IL_0032:  brfalse.s  IL_0036
+  IL_0034:  br.s       IL_0038
+  IL_0036:  br.s       IL_0041
+  IL_0038:  ldloc.1
+  IL_0039:  stloc.3
+  IL_003a:  br.s       IL_003c
+  IL_003c:  ldloc.3
+  IL_003d:  stloc.s    V_7
+  IL_003f:  br.s       IL_004f
+  IL_0041:  ldloca.s   V_6
+  IL_0043:  initobj    ""T""
+  IL_0049:  ldloc.s    V_6
+  IL_004b:  stloc.s    V_7
+  IL_004d:  br.s       IL_004f
+  IL_004f:  ldloc.s    V_7
+  IL_0051:  ret
+}"
+            );
+        }
+
+        [Fact, WorkItem(16195, "https://github.com/dotnet/roslyn/issues/16195")]
+        public void TestMatchWithTypeParameter_02()
+        {
+            var source =
+@"using System;
+class Program
+{
+    public static void Main(string[] args)
+    {
+        Console.Write(M1(2));
+        Console.Write(M2(3));
+        Console.Write(M1(1.1));
+        Console.Write(M2(1.1));
+    }
+    public static int M1<T>(T o)
+    {
+        return o is int t ? t : default(int);
+    }
+    public static int M2<T>(T o)
+    {
+        switch (o)
+        {
+            case int t:
+                return t;
+            default:
+                return default(int);
+        }
+    }
+}";
+            CreateStandardCompilation(source).VerifyDiagnostics(
```

Comment:
>CreateStandardCompilation(source).VerifyDiagnostics( [](start = 12, length = 52)

Consider explicitly using C# 7 version, otherwise this test will break once we bump the default version.